test/adaptive_threshold.jl

```diff
@@ -1,29 +1,94 @@
-
+using ImageBinarization: default_AdaptiveThreshold_window_size
 @testset "adaptive_threshold" begin
-    original_image = testimage("lena")
-    for T in (Gray{N0f8}, Gray{N0f16}, Gray{Float32}, Gray{Float64})
-        img = T.(original_image)
-        img₀₁ = binarize(AdaptiveThreshold(), img)
+    @info "Test: AdaptiveThreshold"
+
+    @testset "API" begin
+        img_gray = imresize(testimage("lena_gray_256"); ratio=0.25)
+        img = copy(img_gray)
+
+        # AdaptiveThreshold
+        @test AdaptiveThreshold() == AdaptiveThreshold(15)
+        @test AdaptiveThreshold(15) == AdaptiveThreshold(percentage=15)
+
+        # window_size non-positive integer
+        f = AdaptiveThreshold()
+        @test_throws ArgumentError binarize(img, f, window_size = -10)
+        @test_throws TypeError binarize(img, f, window_size = 32.5)
+        # percentage ∈ [0, 100]
+        @test_throws ArgumentError AdaptiveThreshold(-10)
+        @test_throws ArgumentError AdaptiveThreshold(150)
+
+        # binarize
+        f = AdaptiveThreshold(percentage=15)
+        binarized_img_1 = binarize(img, f)
+        @test img == img_gray # img unchanged
+        @test eltype(binarized_img_1) == Gray{N0f8}
+        @test binarize(img, f,
+                       window_size=default_AdaptiveThreshold_window_size(img)) == binarized_img_1
 
-        # Check original image is unchanged.
-        @test img == T.(testimage("lena"))
+        binarized_img_2 = binarize(Gray{Bool}, img, f)
+        @test img == img_gray # img unchanged
+        @test eltype(binarized_img_2) == Gray{Bool}
+
+        binarized_img_3 = similar(img, Bool)
+        binarize!(binarized_img_3, img, f)
+        @test img == img_gray # img unchanged
+        @test eltype(binarized_img_3) == Bool
+
+        binarized_img_4 = copy(img_gray)
+        binarize!(binarized_img_4, f)
+        @test eltype(binarized_img_4) == Gray{N0f8}
+
+        @test binarized_img_1 == binarized_img_2
+        @test binarized_img_1 == binarized_img_3
+        @test binarized_img_1 == binarized_img_4
+    end
 
+    @testset "Types" begin
+        # Gray
+        img_gray = imresize(float64.(testimage("lena_gray_256")); ratio=0.25)
+        f = AdaptiveThreshold(percentage=15)
+
+        type_list = generate_test_types([Float32, N0f8], [Gray])
+        for T in type_list
+            img = T.(img_gray)
+            @test_reference "References/AdaptiveThreshold_Gray.png" Gray.(binarize(img, f))
```

Comment:
I didn't know there was such a thing as  @test_reference. Its a great macro! However, I can't see where you added `using ReferenceTests` and if you have added it the package the Project.toml ?

Reply:
This is a PR to the `api` branch, and I've updated some codes there so it's not shown in diff.

`ReferenceTests` is one of them.